pyard/data_repository.py

```diff
@@ -304,7 +318,39 @@ def generate_alleles_and_xx_codes_and_who(db_connection: sqlite3.Connection, img
     db.save_dict(db_connection, 'who_group', flat_who_group,
                  columns=('who', 'allele_list'))
 
-    return valid_alleles, who_alleles, xx_codes, who_group
+    # shortnulls
+    # scan WHO alleles for those with expression characters and make shortnull mappings 
+    # DRB4*01:03N | DRB4*01:03:01:02N/DRB4*01:03:01:13N 
+    # DRB5*01:08N | DRB5*01:08:01N/DRB5*01:08:02N 
+    shortnulls = dict()
+    for k in who_group:
+        if k[-1] not in expression_chars and k[-1] not in ['G', 'P'] and ":" in k:
+            continue
+        # e.g. DRB4*01:03
+        expression_alleles=[]
+        expression_chars_found = set()
+        for an_allele in who_group[k]:
+            # if an allele in a who_group has an expression character but the group allele doesnt, 
+            # add it to shortnulls
+            if an_allele[-1] in expression_chars:
```

Comment:
Here's what I was thinking of.
```
            exp_char = an_allele[-1]
            if exp_char in expression_chars:
                # e.g. DRB4*01:03:01:02N
                expression_chars_found.add(exp_char)
```
A set has unique elements so you don't need to check for membership of an element in a set, just add to it.